flink-ml-lib/src/main/java/org/apache/flink/ml/feature/standardscaler/OnlineStandardScalerModel.java

```diff
@@ -0,0 +1,298 @@
+                                                            getExecutionConfig()))));
+
+            modelDataState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<>(
+                                            "modelData",
+                                            TypeInformation.of(StandardScalerModelData.class)));
+        }
+
+        @Override
+        public void snapshotState(StateSnapshotContext context) throws Exception {
+            super.snapshotState(context);
+            if (modelData != null) {
+                modelDataState.clear();
+                modelDataState.add(modelData);
+            }
+        }
+
+        @Override
+        public void open() throws Exception {
+            super.open();
+            MetricGroup mlModelMetricGroup =
+                    getRuntimeContext()
+                            .getMetricGroup()
+                            .addGroup(MLMetrics.ML_GROUP)
+                            .addGroup(MLMetrics.ML_MODEL_GROUP);
+            mlModelMetricGroup.gauge(MLMetrics.TIMESTAMP, (Gauge<Long>) () -> modelTimeStamp);
+            mlModelMetricGroup.gauge(MLMetrics.VERSION, (Gauge<Long>) () -> modelVersion);
+        }
+
+        @Override
+        public void processElement1(StreamRecord<Row> dataPoint) throws Exception {
+            if (dataPoint.getTimestamp() <= modelTimeStamp + maxAllowedModelDelayMs
+                    && mean != null) {
+                doPrediction(dataPoint);
+            } else {
+                bufferedPointsState.add(dataPoint);
+            }
+        }
+
+        @Override
+        public void processElement2(StreamRecord<StandardScalerModelData> streamRecord)
+                throws Exception {
+            modelData = streamRecord.getValue();
+
+            modelTimeStamp = modelData.timestamp;
+            modelVersion = modelData.version;
+            mean = modelData.mean;
+            DenseVector std = modelData.std;
+
+            if (withStd) {
+                scale = std;
+                double[] scaleValues = scale.values;
+                for (int i = 0; i < scaleValues.length; i++) {
+                    scaleValues[i] = scaleValues[i] == 0 ? 0 : 1 / scaleValues[i];
+                }
+            }
+
+            // Does prediction on the cached data.
+            List<StreamRecord> unprocessedElements = new ArrayList<>();
+            for (StreamRecord dataPoint : bufferedPointsState.get()) {
+                if (dataPoint.getTimestamp() <= modelTimeStamp + maxAllowedModelDelayMs) {
+                    doPrediction(dataPoint);
+                } else {
+                    unprocessedElements.add(dataPoint);
+                }
+            }
+            bufferedPointsState.clear();
+            if (unprocessedElements.size() > 0) {
+                bufferedPointsState.update(unprocessedElements);
```

Comment:
As a minor optimization, can we only clear/update the state if at least one element is predicted in the above for loop?